File: src/runtime/boxes/BbAbort.java
/**
 * BbAbort.java — ABORT: always ω on both ports — force match failure entirely
 * Port of bb_abort.c / bb_abort.s
 *
 * Unlike FAIL (which merely backtracks), ABORT signals the executor to
 * abandon the entire match and take the :F branch immediately.
 * The executor checks for this via the AbortException.
 */
class BbAbort extends BbBox {

    /** Thrown by BbAbort to signal immediate match termination */
    public static final class AbortException extends RuntimeException {
        public AbortException() { super(null, null, true, false); }
    }

    public BbAbort(MatchState ms) { super(ms); }

    @Override public Spec alpha() { throw new AbortException(); }
    @Override public Spec beta()  { throw new AbortException(); }
}

File: src/runtime/boxes/bb_abort.c
/* _XABRT    ABORT       always ω — force match failure */
#include "bb_box.h"
#include <stdlib.h>
#include <string.h>


spec_t bb_abort(void *zeta, int entry)
{
    (void)zeta; (void)entry;
    if (entry==α)                                                               goto ABORT_α;
    if (entry==β)                                                               goto ABORT_β;
    ABORT_α:                                                                    goto ABORT_ω;
    ABORT_β:                                                                    goto ABORT_ω;
    ABORT_ω:                                                                    return spec_empty;
}

abort_t *bb_abort_new(void)
{ return calloc(1,sizeof(abort_t)); }
